### PR TITLE
Refactor with LibmeshPetscCall

### PR DESCRIPTION
**Pull request by nmnobre**
Hi,

Would this be of interest? I can probably do the rest of the code in a morning if it would.

Cheers,
-N

### CODE REVIEW

**Comment by jwpeterson** [comment]
I assume that it was @lindsayad's intention to eventually replace all of the PETSc function call/error checking code with `LibmeshPetscCall`, so this looks like a good change to me, but I'll let him comment whether there is some reason to hold off on updating the remaining ones.

@roystgnr this PR was marked as a Draft on GitHub but the CI testing still ran. Is that the intended behavior in CIVET for Draft PRs? I seem to remember that marking a PR as Draft would prevent the tests from running before the user was ready to do so.

**Comment by nmnobre** [comment]
> I assume that it was @lindsayad's intention to eventually replace all of the PETSc function call/error checking code with `LibmeshPetscCall`, so this looks like a good change to me, but I'll let him comment whether there is some reason to hold off on updating the remaining ones.

I think I've completed all replacements for the macros we already have.
Unfortunately, I think we should have named `LibmeshPetscCall(2)`, `LibmeshPetscCallAbort(2)` instead, so that `LibmeshPetscCall` would be free for idioms using `CHKERRQ`. As a workaround, I can create `LibmeshPetscCall3` for the latter case and redefine `LibmeshPetscCall2` to get an MPI communicator, so I can cover more cases.
This way, when we think we're ready to drop support for PETSc < 3.17.0, we could simply:
- redefine `LibmeshPetscCall` to wrap `PetscCallAbort`;
- drop `LibmeshPetscCall2` and call `PetscCallAbort` directly;
- drop `LibmeshPetscCall3` and call `PetscCall` directly.

> @roystgnr this PR was marked as a Draft on GitHub but the CI testing still ran. Is that the intended behavior in CIVET for Draft PRs? I seem to remember that marking a PR as Draft would prevent the tests from running before the user was ready to do so.

FWIW, I actually find this very useful. :)

**Comment by roystgnr** [comment]
Running CI for Draft PRs is definitely the behavior I'd intend.  I write ~~too much~~ occasional code that I want to run through the whole CI gauntlet even when I know there are other things that still need to be improved or even fixed before merging.

**Comment by roystgnr** [comment]
I'd rather not use `PetscCall` ever, personally.  `LIBMESH_CHKERR` doesn't avoid using `CHKERR` not because the latter is a recent invention, but because throwing an exception that can either be caught or can trigger a terminate handler is strictly more useful than directly triggering a terminate handler.

**Comment by nmnobre** [comment]
> I'd rather not use `PetscCall` ever, personally. `LIBMESH_CHKERR` doesn't avoid using `CHKERR` not because the latter is a recent invention, but because throwing an exception that can either be caught or can trigger a terminate handler is strictly more useful than directly triggering a terminate handler.

Yes, of course. Brain fart -.-, I was looking through the non-`LIBMESH_ENABLE_EXCEPTIONS` code path for a few hours to make sure I was calling the right macro, and completely forgot about the default code path. I think tomorrow I'll create `LibmeshPetscCallQ` which I think is a bit more descriptive than what I suggested earlier with `LibmeshPetscCall3`. (I think you guys might've named it `LibmeshPetscCall2` because it takes 2 args).

**Comment by nmnobre** [comment]
Done. All yours @roystgnr.

File: include/base/libmesh.h
Comment (jwpeterson) [inline include/base/libmesh.h:122]: I'd rather not be throwing exceptions from the `LibMeshInit` destructor (or any destructor for that matter) since it's generally considered bad practice (destructors are called during stack unwinding, throwing another exception during stack unwinding immediately terminates the program). We use `libmesh_exceptionless_assert()` and `CHKERRABORT` in the `LibMeshInit` destructor for this reason.

Comment (nmnobre) [reply include/base/libmesh.h]: Agreed.

Comment (nmnobre) [reply include/base/libmesh.h]: I chose to just revert to using `CHKERRABORT` instead of calling `PetscCallAbort` for compatibility with PETSc < 3.17.0.

**Comment by lindsayad** [comment]
You can use `[WIP]` or `WIP:` in the PR title for disabling CIVET testing

File: include/numerics/petsc_solver_exception.h
Comment (lindsayad) [inline include/numerics/petsc_solver_exception.h:157]: With all these macros now, please add comments about when developers should use which

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Review by lindsayad — APPROVED** [review]

File: include/numerics/petsc_solver_exception.h
Comment (lindsayad) [inline include/numerics/petsc_solver_exception.h:94]: It's confusing that the one vs. two argument macros are equivalent to `PetscCall` and `PetscCallAbort` respectively. Is this only because MOOSE is "being dumb"?

Comment (nmnobre) [reply include/numerics/petsc_solver_exception.h:94]: Sorry, I don't think I understood the question, `PetscCall` and `PetscCallAbort` are also one and two argument macros?

Comment (lindsayad) [reply include/numerics/petsc_solver_exception.h:94]: `PetscCallAbort` will abort immediately if the error code is non-zero. This does not happen with `PetscCall`

Comment (nmnobre) [reply include/numerics/petsc_solver_exception.h:94]: Agreed. Point being? (sorry, it's late over here...)

Comment (lindsayad) [reply include/numerics/petsc_solver_exception.h:94]: I would not expect that `LIBMESH_CHKERR` and `LIBMESH_CHKERR2` have fundamentally different behavior when the only difference is that one takes a comm and the other does not

Comment (nmnobre) [reply include/numerics/petsc_solver_exception.h:94]: Right, thanks, apologies. I'm with you now.

Yes, it's true that before, *if* `LIBMESH_ENABLE_EXCEPTIONS` wasn't defined, they'd both wind up calling `CHKERRABORT`, and that behaviour is now reserved to `LIBMESH_CHKERRA`. Note, however, that the fallback is only defined for the case where `LIBMESH_ENABLE_EXCEPTIONS` is indeed defined, for which the behaviour was already the same before.

Comment (lindsayad) [reply include/numerics/petsc_solver_exception.h:94]: Got ya. Yea I thought you didn't likely *add* any silliness that wasn't there before. Still would be nice if we as a collective make this more consistent. Probably can do this after updating MOOSE

**Pull request merged**
